tests/bluetooth/tester/src/btp_cap.c

```diff
@@ -0,0 +1,1053 @@
+							BTP_CAP_UNICAST_START_STATUS_FAILED);
+
+		return;
+	}
+
+	btp_send_cap_unicast_start_completed_ev(group->index,
+						BTP_CAP_UNICAST_START_STATUS_SUCCESS);
+}
+
+static void unicast_update_complete_cb(int err, struct bt_conn *conn)
+{
+	LOG_DBG("");
+
+	if (err != 0) {
+		LOG_DBG("Failed to unicast-update, err %d", err);
+	}
+}
+
+static void unicast_stop_complete_cb(struct bt_bap_unicast_group *group, int err,
+				     struct bt_conn *conn)
+{
+	LOG_DBG("");
+
+	if (err != 0) {
+		LOG_DBG("Failed to unicast-stop, err %d", err);
+		btp_send_cap_unicast_stop_completed_ev(group->index,
+						       BTP_CAP_UNICAST_START_STATUS_FAILED);
+
+		return;
+	}
+
+	btp_send_cap_unicast_stop_completed_ev(group->index,
+					       BTP_CAP_UNICAST_START_STATUS_SUCCESS);
+}
+
+static struct bt_cap_initiator_cb cap_cb = {
+	.unicast_discovery_complete = cap_discovery_complete_cb,
+	.unicast_start_complete = unicast_start_complete_cb,
+	.unicast_update_complete = unicast_update_complete_cb,
+	.unicast_stop_complete = unicast_stop_complete_cb,
+};
+
+static uint8_t btp_cap_supported_commands(const void *cmd, uint16_t cmd_len,
+					  void *rsp, uint16_t *rsp_len)
+{
+	struct btp_cap_read_supported_commands_rp *rp = rsp;
+
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_CAP_READ_SUPPORTED_COMMANDS);
+	tester_set_bit(rp->data, BTP_CAP_DISCOVER);
+
+	*rsp_len = sizeof(*rp) + 1;
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static uint8_t btp_cap_discover(const void *cmd, uint16_t cmd_len,
+				void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_cap_discover_cmd *cp = cmd;
+	struct bt_conn *conn;
+	int err;
+
+	LOG_DBG("");
+
+	conn = bt_conn_lookup_addr_le(BT_ID_DEFAULT, &cp->address);
+	if (!conn) {
+		LOG_ERR("Unknown connection");
+		return BTP_STATUS_FAILED;
+	}
+
+	err = bt_cap_initiator_unicast_discover(conn);
+	if (err != 0) {
+		LOG_DBG("Failed to discover remote ASEs: %d", err);
+		bt_conn_unref(conn);
+
+		return BTP_STATUS_FAILED;
+	}
+
+	bt_conn_unref(conn);
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static void csip_set_coordinator_lock_set_cb(int err)
+{
+	LOG_DBG("");
+}
+
+static void csip_set_coordinator_lock_release_cb(int err)
+{
+	LOG_DBG("");
+}
+
+static void csip_discover_cb(struct bt_conn *conn,
+			     const struct bt_csip_set_coordinator_set_member *member,
+			     int err, size_t set_count)
+{
+	LOG_DBG("");
+
+	uint8_t conn_index;
+
+	if (err != 0) {
+		LOG_DBG("discover failed (%d)", err);
+		return;
+	}
+
+	if (set_count == 0) {
+		LOG_DBG("Device has no sets");
+		return;
+	}
+
+	conn_index = bt_conn_index(conn);
+
+	LOG_DBG("Found %zu sets on member[%u]", set_count, conn_index);
+
+	for (size_t i = 0U; i < set_count; i++) {
+		LOG_DBG("CSIS[%zu]: %p", i, &member->insts[i]);
+		LOG_DBG("Rank: %u", member->insts[i].info.rank);
+		LOG_DBG("Set Size: %u", member->insts[i].info.set_size);
+		LOG_DBG("Lockable: %u", member->insts[i].info.lockable);
+	}
+
+	cur_csis_inst = &((struct bt_csip_set_coordinator_set_member *)member)->insts[0];
```

Comment:
```suggestion
	cur_csis_inst = &member->insts[0];
```

Reply:
This one is needed to fix a warning about implicit discard of const qualifier. The declaration of the callback:
https://github.com/zephyrproject-rtos/zephyr/blob/c3e8b731eff5beb5b28aef248bbc8ed20a2c7419/include/zephyr/bluetooth/audio/csip.h#L279-L282C29

Reply:
IMO it is a poor design to cast away `const`. If something is marked `const`, there usually is a reason for it. 

If you need to store a non-const version of it, you should `memcpy` it. 

That being said, I don't see why `cur_csis_inst` couldn't be `const`. It seems like it is only being used for `bt_csip_set_coordinator_ordered_access` where it is also used as a `const`.

Reply:
I agree. I mislooked that `const` near set_info somehow, thank you.